pushsync/pusher.go

```diff
@@ -0,0 +1,343 @@
+			if unsubscribe != nil {
+				unsubscribe()
+			}
+
+			log.Debug("set chunk status to synced, insert to db GC index")
+			syncedTags := make(map[uint32]int)
+			// set chunk status to synced, insert to db GC index
+			if err := p.store.Set(ctx, chunk.ModeSetSync, syncedAddrs...); err != nil {
+				log.Error("error setting chunks to synced", "err", err)
+				continue
+			}
+			for i, item := range syncedItems {
+				// increment synced count for the tag if exists
+				tag := item.tag
+				if tag != nil {
+					syncedTags[tag.Uid] = syncedTags[tag.Uid] + 1
+					item.span.Finish()
+				}
+				delete(p.pushed, syncedAddrs[i].Hex())
+			}
+			// iterate over tags in this batch
+			for uid, n := range syncedTags {
+				tag, _ := p.tags.Get(uid)
+				tag.IncN(chunk.StateSynced, n)
+				if tag.Done(chunk.StateSynced) {
+					log.Info("closing root span for tag", "taguid", tag.Uid, "tagname", tag.Name)
+					tag.FinishRootSpan()
+				}
+			}
+			// reset synced list
+			syncedAddrs = nil
+			syncedItems = nil
+
+			// we don't want to record the first iteration
+			if chunksInBatch != -1 {
+				// this measurement is not a timer, but we want a histogram, so it fits the data structure
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.hist", nil).Update(time.Duration(chunksInBatch))
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.time", nil).UpdateSince(batchStartTime)
+				metrics.GetOrRegisterCounter("pusher.subscribe-push.chunks-in-batch", nil).Inc(int64(chunksInBatch))
+			}
+			chunksInBatch = 0
+			batchStartTime = time.Now()
+
+			// and start iterating on Push index from the beginning
+			chunks, unsubscribe = p.store.SubscribePush(ctx)
+			// reset timer to go off after retryInterval
+			timer.Reset(retryInterval)
+
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			// if no more, set to nil, reset timer to 0 to finalise batch immediately
+			if !more {
+				chunks = nil
+				timer.Reset(0)
+				break
+			}
+
+			chunksInBatch++
+			metrics.GetOrRegisterCounter("pusher.send-chunk", nil).Inc(1)
+			// if no need to sync this chunk then continue
+			if !p.needToSync(ch) {
+				break
+			}
+
+			metrics.GetOrRegisterCounter("pusher.send-chunk.send-to-sync", nil).Inc(1)
+			// send the chunk and ignore the error
+			// go func(ch chunk.Chunk) {
+			if err := p.sendChunkMsg(ch); err != nil {
+				log.Error("error sending chunk", "addr", ch.Address().Hex(), "err", err)
+			}
+
+		// handle incoming receipts
+		case addr := <-p.receipts:
+			hexaddr := hex.EncodeToString(addr)
+			log.Debug("synced", "addr", hexaddr)
+			metrics.GetOrRegisterCounter("pusher.receipts.all", nil).Inc(1)
+			// ignore if already received receipt
+			item, found := p.pushed[hexaddr]
+			if !found {
+				metrics.GetOrRegisterCounter("pusher.receipts.not-found", nil).Inc(1)
+				log.Debug("not wanted or already got... ignore", "addr", hexaddr)
+				break
+			}
+			if item.synced {
```

Comment:
does this denote a case where we've received two receipts for the same chunk? please comment